contracts/voting-verifier/src/contract.rs

```diff
@@ -64,3 +65,781 @@ pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+            deps.as_mut(),
+            mock_env(),
+            mock_info(SENDER, &[]),
+            ExecuteMsg::VerifyWorkerSet {
+                message_id: message_id("id", 0),
+                new_operators: operators.clone(),
+            },
+        );
+        assert!(res.is_ok());
+
+        for worker in workers {
+            let res = execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(worker.address.as_str(), &[]),
+                ExecuteMsg::Vote {
+                    poll_id: 2u64.into(),
+                    votes: vec![Vote::SucceededOnChain],
+                },
+            );
+            assert!(res.is_ok());
+        }
+
+        let res = execute(
+            deps.as_mut(),
+            mock_env_expired(),
+            mock_info(SENDER, &[]),
+            ExecuteMsg::EndPoll {
+                poll_id: 2u64.into(),
+            },
+        );
+        assert!(res.is_ok());
+
+        let res: VerificationStatus = from_binary(
+            &query(
+                deps.as_ref(),
+                mock_env(),
+                QueryMsg::GetWorkerSetStatus {
+                    new_operators: operators.clone(),
+                },
+            )
+            .unwrap(),
+        )
+        .unwrap();
+        assert_eq!(res, VerificationStatus::SucceededOnChain);
+    }
+
+    #[test]
+    fn should_not_confirm_twice() {
+        let mut deps = setup();
+
+        let operators = Operators::new(vec![(vec![0, 1, 0, 1].into(), 1u64.into())], 1u64.into());
+        let res = execute(
+            deps.as_mut(),
+            mock_env(),
+            mock_info(SENDER, &[]),
+            ExecuteMsg::VerifyWorkerSet {
+                message_id: message_id("id", 0),
+                new_operators: operators.clone(),
+            },
+        );
+        assert!(res.is_ok());
+
+        for worker in workers() {
+            let res = execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(worker.address.as_str(), &[]),
+                ExecuteMsg::Vote {
+                    poll_id: 1u64.into(),
+                    votes: vec![Vote::SucceededOnChain],
+                },
+            );
+            assert!(res.is_ok());
+        }
+
+        let res = execute(
+            deps.as_mut(),
+            mock_env_expired(),
+            mock_info(SENDER, &[]),
+            ExecuteMsg::EndPoll {
+                poll_id: 1u64.into(),
+            },
+        );
+        assert!(res.is_ok());
+
+        // try again, should fail
+        let err = execute(
+            deps.as_mut(),
+            mock_env_expired(),
+            mock_info(SENDER, &[]),
+            ExecuteMsg::VerifyWorkerSet {
+                message_id: message_id("id", 0),
+                new_operators: operators.clone(),
+            },
+        )
+        .unwrap_err();
+        are_contract_err_strings_equal(err, ContractError::WorkerSetAlreadyConfirmed);
+    }
```

Comment:
this also seems inconsistent between messages and worker sets, messages will just return the confirmed state without triggering a new verification, do we need to resolve this?